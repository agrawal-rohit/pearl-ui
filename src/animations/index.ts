import { useAnimationState } from "../hooks/useAnimationState";

/**
 * Custom hook for creating a fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from an opacity of 0 to an opacity of 1.
 * This creates a fade-in effect.
 *
 * @returns An animation state that creates a fade-in effect.
 */
export const useFade = () => {
  // Define the initial and final states for the fade animation.
  // The animation will start with an opacity of 0 (completely transparent) and end with an opacity of 1 (completely opaque).
  const fadeAnimationState = useAnimationState({
    from: {
      opacity: 0,
    },
    to: {
      opacity: 1,
    },
  });

  // Return the fade animation state.
  return fadeAnimationState;
};

/**
 * Custom hook for creating a scale and fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a scale of 0.5 and opacity of 0 to a scale of 1 and opacity of 1.
 * This creates a combined effect of scaling and fading in.
 *
 * @returns An animation state that creates a scale and fade-in effect.
 */
export const useScaleFade = () => {
  // Define the initial and final states for the scale and fade animation.
  // The animation will start with a scale of 0.5 (half the original size) and opacity of 0 (completely transparent)
  // and end with a scale of 1 (original size) and opacity of 1 (completely opaque).
  const scaleFadeAnimationState = useAnimationState({
    from: {
      scale: 0.9,
      opacity: 0,
    },
    to: {
      scale: 1,
      opacity: 1,
    },
  });

  // Return the scale and fade animation state.
  return scaleFadeAnimationState;
};

/**
 * Custom hook for creating a slide animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a translateX of -100 to a translateX of 0.
 * This creates a slide-in effect.
 *
 * @returns An animation state that creates a slide-in effect.
 */
export const useSlide = () => {
  // Define the initial and final states for the slide animation.
  // The animation will start with a translateX of -100 (completely off-screen to the left) and end with a translateX of 0 (original position).
  const slideAnimationState = useAnimationState({
    from: {
      translateY: -100,
    },
    to: {
      translateY: 0,
    },
  });

  // Return the slide animation state.
  return slideAnimationState;
};

/**
 * Custom hook for creating a slide and fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a translateX of -100 and opacity of 0 to a translateX of 0 and opacity of 1.
 * This creates a combined effect of sliding and fading in.
 *
 * @returns An animation state that creates a slide and fade-in effect.
 */
export const useSlideFade = () => {
  // Define the initial and final states for the slide and fade animation.
  // The animation will start with a translateX of -100 (completely off-screen to the left) and opacity of 0 (completely transparent)
  // and end with a translateX of 0 (original position) and opacity of 1 (completely opaque).
  const slideFadeAnimationState = useAnimationState({
    from: {
      opacity: 0,
      translateY: 10,
    },
    to: {
      opacity: 1,
      translateY: 0,
    },
  });

  // Return the slide and fade animation state.
  return slideFadeAnimationState;
};

/**
 * Custom hook for creating a collapse animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a scaleY of 1 to a scaleY of 0.
 * This creates a collapse effect.
 *
 * @returns An animation state that creates a collapse effect.
 */
export const useCollapse = () => {
  // Define the initial and final states for the collapse animation.
  // The animation will start with a scaleY of 1 (original size) and end with a scaleY of 0 (collapsed).
  const collapseAnimationState = useAnimationState({
    from: {
      opacity: 0,
      height: 0,
    },
    to: {
      opacity: 1,
      height: 100,
    },
  });

  // Return the collapse animation state.
  return collapseAnimationState;
};
