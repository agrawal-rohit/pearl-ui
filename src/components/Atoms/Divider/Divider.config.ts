export default {
  baseStyle: {
    bg: {
      light: "neutral.300",
      dark: "neutral.900",
    },
    thickness: 1,
    length: "100%",
  },
  defaults: {},
};
