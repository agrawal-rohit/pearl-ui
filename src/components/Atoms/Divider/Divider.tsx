import React from "react";
import { View } from "react-native";
import Box, { BoxProps } from "../../Atoms/Box/Box";
import {
  ComponentSizes,
  ComponentVariants,
  ResponsiveValue,
} from "../../../theme/src/types";
import { pearlify } from "../../../hooks/pearlify";

export type DividerProps = BoxProps & {
  /** The size of the divider */
  size?: ResponsiveValue<ComponentSizes<"Divider">>;
  /** The variant of the divider */
  variant?: ResponsiveValue<ComponentVariants<"Divider">>;
  /** The length of the divider */
  length?: number | string;
  /** The thickness of the divider */
  thickness?: number;
  /** The direction of the divider */
  orientation?: "horizontal" | "vertical";
};

const CustomDivider: React.FC<DividerProps> = ({ children, ...props }) => {
  return (
    <Box
      style={{
        ...(props.style as any),
        height:
          props.orientation === "horizontal" ? props.thickness : props.length,
        width:
          props.orientation === "vertical" ? props.thickness : props.length,
      }}
    >
      {children}
    </Box>
  );
};

/** Divider is used to visually separate content in a list or group. */
const Divider = pearlify(CustomDivider, {
  componentName: "Divider",
  type: "atom",
});

export default Divider;
