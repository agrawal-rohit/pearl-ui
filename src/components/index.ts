// Atoms
export { default as Box, BoxProps } from "./atoms/box/box";
export { default as Center } from "./atoms/center/center";
export { default as Collapse, CollapseProps } from "./atoms/collapse/collapse";
export { default as Divider, DividerProps } from "./atoms/divider/divider";
export { default as Fade, FadeProps } from "./atoms/fade/fade";
export { default as Grid, GridProps } from "./atoms/grid/grid";
export { default as Icon, IconProps } from "./atoms/icon/icon";
export {
  default as Pressable,
  PressableProps,
} from "./atoms/pressable/pressable";
export {
  default as ScaleFade,
  ScaleFadeProps,
} from "./atoms/scale-fade/scale-fade";
export {
  default as List,
  OrderedList,
  UnorderedList,
  ListProps,
} from "./atoms/list/list";
export { default as Screen, ScreenProps } from "./atoms/screen/screen";
export { default as Skeleton, SkeletonProps } from "./atoms/skeleton/skeleton";
export {
  default as SkeletonCircle,
  SkeletonCircleProps,
} from "./atoms/skeleton/skeleton-circle";
export {
  default as SkeletonText,
  SkeletonTextProps,
} from "./atoms/skeleton/skeleton-text";
export { default as Slide, SlideProps } from "./atoms/slide/slide";
export {
  default as SlideFade,
  SlideFadeProps,
} from "./atoms/slide-fade/slide-fade";
export { default as Spacer } from "./atoms/spacer/spacer";
export { default as Spinner, SpinnerProps } from "./atoms/spinner/spinner";
export {
  default as Stack,
  HStack,
  VStack,
  ZStack,
  HStackProps,
  VStackProps,
  StackProps,
} from "./atoms/stack/stack";
export { default as Text, TextProps } from "./atoms/text/text";

// Molecules
export { default as Avatar, AvatarProps } from "./molecules/avatar/avatar";
export {
  default as AvatarGroup,
  AvatarGroupProps,
} from "./molecules/avatar/avatar-group";
export { default as Badge, BadgeProps } from "./molecules/badge/badge";
export {
  default as withBadge,
  WithBadgeOptions,
} from "./molecules/badge/withBadge";
export { default as Button, ButtonProps } from "./molecules/button/button";
export {
  default as ButtonGroup,
  ButtonGroupProps,
} from "./molecules/button/button-group";
export {
  default as CheckBox,
  CheckBoxProps,
} from "./molecules/checkbox/checkbox";
export {
  default as CheckBoxGroup,
  CheckBoxGroupProps,
} from "./molecules/checkbox/checkbox-group";
export {
  default as IconButton,
  IconButtonProps,
} from "./molecules/icon-button/icon-button";
export { default as Image, ImageProps } from "./molecules/image/image";
export { default as Input, InputProps } from "./molecules/input/input";
export {
  default as PinInput,
  PinInputProps,
} from "./molecules/pin-input/pin-input";
export {
  default as Progress,
  ProgressProps,
} from "./molecules/progress/progress";
export { default as Radio, RadioProps } from "./molecules/radio/radio";
export {
  default as RadioGroup,
  RadioGroupProps,
} from "./molecules/radio/radio-group";
export { default as Switch, SwitchProps } from "./molecules/switch/switch";
export {
  default as Textarea,
  TextareaProps,
} from "./molecules/textarea/textarea";
export { default as Video, VideoProps } from "./molecules/video/video";
export {
  default as Accordion,
  AccordionProps,
  AccordionItemProps,
  AccordionButtonProps,
  AccordionPanelProps,
  AccordionItem,
  AccordionButton,
  AccordionPanel,
} from "./molecules/accordion/accordion";
export { default as Alert, AlertProps } from "./molecules/alert/alert";
