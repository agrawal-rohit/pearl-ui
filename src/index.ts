export { default as Box } from "./components/Box/Box";
export { default as Button } from "./components/Button/Button";
export { default as Text } from "./components/Text/Text";
