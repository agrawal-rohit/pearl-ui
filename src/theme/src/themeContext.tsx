import AppLoading from "expo-app-loading";
import {
  useFonts,
  Poppins_300Light,
  Poppins_400Regular,
  Poppins_500Medium,
  Poppins_600SemiBold,
  Poppins_700Bold,
  Poppins_800ExtraBold,
} from "@expo-google-fonts/poppins";
import React, { useEffect, useState, createContext } from "react";
import { useColorScheme } from "react-native";
import { baseTheme } from "./basetheme";
import { BasePearlTheme } from "./types";

type ThemeType = "light" | "dark" | "system";

export interface IThemeContext {
  /** Theme configuration object for the active color mode */
  theme: BasePearlTheme;
  /** Active color mode */
  colorMode: ThemeType;
  /** Function to toggle the active color mode */
  toggleColorMode(): void;
}

interface ThemeProviderProps {
  /** Initial color mode for the theme (light, dark, system) */
  initialColorMode?: ThemeType;
  /** The theme configuration object */
  theme?: BasePearlTheme;
  /** A mapping object of the fonts used by the theme */
  fonts?: object;
  /**React children */
  children: React.ReactNode;
}

const baseFonts = {
  "Poppins-Light": Poppins_300Light,
  "Poppins-Regular": Poppins_400Regular,
  "Poppins-Medium": Poppins_500Medium,
  "Poppins-SemiBold": Poppins_600SemiBold,
  "Poppins-Bold": Poppins_700Bold,
  "Poppins-ExtraBold": Poppins_800ExtraBold,
};

export const themeContext = createContext({} as IThemeContext);

/**
 * The main provider component which controls the theme and color mode of the components in the entire application.
 */
export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  initialColorMode = "light",
  theme = baseTheme,
  fonts = baseFonts,
  children,
}) => {
  const [colorMode, setColorMode] = useState<ThemeType>(initialColorMode);
  const systemThemeStyle = useColorScheme() as ThemeType;
  const [haveFontsLoaded] = useFonts(fonts);

  const changeThemeTo = (colorMode: ThemeType) => {
    if (colorMode === "dark") {
      setColorMode("dark");
    } else {
      setColorMode("light");
    }
  };

  const toggleColorMode = () => {
    if (colorMode === "light") {
      changeThemeTo("dark");
    } else {
      changeThemeTo("light");
    }
  };

  useEffect(() => {
    // Set initial Theme
    if (colorMode === "light") {
      changeThemeTo("light");
    } else if (colorMode === "dark") {
      changeThemeTo("dark");
    } else if (colorMode === "system") {
      changeThemeTo(systemThemeStyle);
    }
  }, [colorMode, initialColorMode, theme]);

  if (haveFontsLoaded) {
    return (
      <themeContext.Provider
        value={{ theme, colorMode, toggleColorMode } as IThemeContext}
      >
        {children}
      </themeContext.Provider>
    );
  }

  return <AppLoading />;
};
