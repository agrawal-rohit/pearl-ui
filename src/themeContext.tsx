import React, { useEffect, useState, useContext, createContext } from "react";
import { useColorScheme } from "react-native";
import { ThemeProvider as RestyleThemeProvider } from "@shopify/restyle";
import { baseTheme, darkTheme, Theme } from "./theme";
import { ReactElement } from "react-color/node_modules/@types/react";

type ThemeType = "light" | "dark" | "system";

interface IThemeContext {
  themeMode: ThemeType;
  toggleTheme(theme: string): void;
}

interface ThemeProviderProps {
  defaultMode?: ThemeType;
  light?: Theme;
  dark?: Theme;
}

export const themeContext = createContext({} as IThemeContext);

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  defaultMode = "light",
  light = baseTheme,
  dark = darkTheme,
  children,
}) => {
  const [themeMode, setThemeMode] = useState<ThemeType>(defaultMode);
  const systemThemeStyle = useColorScheme() as ThemeType;

  const toggleTheme = () => {
    if (themeMode === "light") {
      setThemeMode("dark");
    } else {
      setThemeMode("light");
    }
  };

  useEffect(() => {
    // Set theme
    if (defaultMode === "system") {
      setThemeMode(systemThemeStyle);
    }
  }, []);

  return (
    <themeContext.Provider value={{ themeMode, toggleTheme } as IThemeContext}>
      <RestyleThemeProvider theme={themeMode === "light" ? light : dark}>
        {children}
      </RestyleThemeProvider>
    </themeContext.Provider>
  );
};

export const useTheme = () => useContext(themeContext);
