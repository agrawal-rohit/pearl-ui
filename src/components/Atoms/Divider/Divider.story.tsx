import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Divider from "./Divider";

storiesOf("Divider", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("horizontal", () => <Divider />)
  .add("vertical", () => <Divider orientation="vertical" />);
