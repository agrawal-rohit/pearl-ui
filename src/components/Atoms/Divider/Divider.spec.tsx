import React from "react";
import { render } from "@testing-library/react-native";
import Divider from "./Divider";
import { ThemeProvider } from "../../../theme/src/themeContext";

jest.useFakeTimers();

describe("Atoms/Divider", () => {
  it("passes the snapshot test in light mode", () => {
    const tree = render(
      <ThemeProvider>
        <Divider orientation="horizontal" />
        <Divider orientation="vertical" />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test in dark mode", () => {
    const tree = render(
      <ThemeProvider initialColorMode="dark">
        <Divider orientation="horizontal" />
        <Divider orientation="vertical" />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test for different length modes", () => {
    const tree = render(
      <ThemeProvider initialColorMode="dark">
        <Divider length="50%" />
        <Divider orientation="vertical" length={20} bgColor="violet" />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
