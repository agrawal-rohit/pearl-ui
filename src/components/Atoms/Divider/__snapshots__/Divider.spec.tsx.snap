// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Divider passes the snapshot test for different length modes 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#1A2138",
        "display": "flex",
        "height": 1,
        "width": "50%",
      }
    }
  />,
  <View
    style={
      Object {
        "backgroundColor": "#ee82ee",
        "display": "flex",
        "height": 20,
        "width": 1,
      }
    }
  />,
]
`;

exports[`Atoms/Divider passes the snapshot test in dark mode 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#1A2138",
        "display": "flex",
        "height": 1,
        "width": "100%",
      }
    }
  />,
  <View
    style={
      Object {
        "backgroundColor": "#1A2138",
        "display": "flex",
        "height": "100%",
        "width": 1,
      }
    }
  />,
]
`;

exports[`Atoms/Divider passes the snapshot test in light mode 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#E4E9F2",
        "display": "flex",
        "height": 1,
        "width": "100%",
      }
    }
  />,
  <View
    style={
      Object {
        "backgroundColor": "#E4E9F2",
        "display": "flex",
        "height": "100%",
        "width": 1,
      }
    }
  />,
]
`;
